In [23]:
import pandas as pd
import numpy as np


In [24]:
url = r"C:\Users\Anih Nathan\OneDrive\Desktop\Hamoye fall 23\stageb\energydata_complete.csv"
quizdata = pd.read_csv(url)
quizdata

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
...                    ...         ...     ...        ...        ...   
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3         T4  ...         T9  \
0      19.200000  44.790000  19.790000  44.730000  19.000000  ...  17.033333   
1      19.200000  44.722500  19.790000  44.790000  19.000000  ...  17.066667   
2      19.200000  44.626667  19.790000  44.933333  18.926667  ...  17.000000   
3      19.200000  44.590000  19.790000  45.000000  18.890000  ...  17.000000   
4      19.200000  44.530000  19.790000  45.000000  18.890000  ...  17.000000   
...          ...        ...        ...        ...        ...  ...        ...   
19730  25.890000  42.025714  27.200000  41.163333  24.700000  ...  23.200000   
19731  25.754000  42.080000  27.133333  41.223333  24.700000  ...  23.200000   
19732  25.628571  42.768571  27.050000  41.690000  24.700000  ...  23.200000   
19733  25.414000  43.036000  26.890000  41.290000  24.700000  ...  23.200000   
19734  25.264286  42.971429  26.823333  41.156667  24.700000  ...  23.200000   

          RH_9      T_out  Press_mm_hg     RH_out  Windspeed  Visibility  \
0      45.5300   6.600000        733.5  92.000000   7.000000   63.000000   
1      45.5600   6.483333        733.6  92.000000   6.666667   59.166667   
2      45.5000   6.366667        733.7  92.000000   6.333333   55.333333   
3      45.4000   6.250000        733.8  92.000000   6.000000   51.500000   
4      45.4000   6.133333        733.9  92.000000   5.666667   47.666667   
...        ...        ...          ...        ...        ...         ...   
19730  46.7900  22.733333        755.2  55.666667   3.333333   23.666667   
19731  46.7900  22.600000        755.2  56.000000   3.500000   24.500000   
19732  46.7900  22.466667        755.2  56.333333   3.666667   25.333333   
19733  46.8175  22.333333        755.2  56.666667   3.833333   26.166667   
19734  46.8450  22.200000        755.2  57.000000   4.000000   27.000000   

       Tdewpoint        rv1        rv2  
0       5.300000  13.275433  13.275433  
1       5.200000  18.606195  18.606195  
2       5.100000  28.642668  28.642668  
3       5.000000  45.410389  45.410389  
4       4.900000  10.084097  10.084097  
...          ...        ...        ...  
19730  13.333333  43.096812  43.096812  
19731  13.300000  49.282940  49.282940  
19732  13.266667  29.199117  29.199117  
19733  13.233333   6.322784   6.322784  
19734  13.200000  34.118851  34.118851  

[19735 rows x 29 columns]

In [25]:
selected_xy = quizdata[['T2','T6']]
selected_xy

T2         T6
0      19.200000   7.026667
1      19.200000   6.833333
2      19.200000   6.560000
3      19.200000   6.433333
4      19.200000   6.366667
...          ...        ...
19730  25.890000  24.796667
19731  25.754000  24.196667
19732  25.628571  23.626667
19733  25.414000  22.433333
19734  25.264286  21.026667

[19735 rows x 2 columns]

In [26]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
xy_model = LinearRegression()

x = np.array(selected_xy['T2']).reshape(-1,1)
y = np.array(selected_xy['T6'])

xy_model.fit(x,y)


LinearRegression()

In [27]:
y_predict = xy_model.predict(x)
mse = mean_squared_error(y,y_predict)
rmse = np.sqrt(mse)
print(round(rmse,3))

3.644


#### Q17 Root Mean Squared error in three D.P is 3.644

In [28]:
quizdata = quizdata.drop(['date','lights'],axis=1)
quizdata.head()

Appliances     T1       RH_1    T2       RH_2     T3       RH_3         T4  \
0          60  19.89  47.596667  19.2  44.790000  19.79  44.730000  19.000000   
1          60  19.89  46.693333  19.2  44.722500  19.79  44.790000  19.000000   
2          50  19.89  46.300000  19.2  44.626667  19.79  44.933333  18.926667   
3          50  19.89  46.066667  19.2  44.590000  19.79  45.000000  18.890000   
4          60  19.89  46.333333  19.2  44.530000  19.79  45.000000  18.890000   

        RH_4         T5  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  45.566667  17.166667  ...  17.033333  45.53  6.600000        733.5    92.0   
1  45.992500  17.166667  ...  17.066667  45.56  6.483333        733.6    92.0   
2  45.890000  17.166667  ...  17.000000  45.50  6.366667        733.7    92.0   
3  45.723333  17.166667  ...  17.000000  45.40  6.250000        733.8    92.0   
4  45.530000  17.200000  ...  17.000000  45.40  6.133333        733.9    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  
2   6.333333   55.333333        5.1  28.642668  28.642668  
3   6.000000   51.500000        5.0  45.410389  45.410389  
4   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 27 columns]

In [29]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

In [30]:

feature_data = quizdata.drop(columns=['Appliances'])
appliance_target = quizdata['Appliances']
X_train,X_test,y_train,y_test = train_test_split(feature_data, appliance_target, test_size=0.3, random_state=42)


In [31]:
scaler = MinMaxScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)
mymodel = LinearRegression()
mymodel.fit(X_train,y_train)
y_trained = mymodel.predict(X_train)
y_tested = mymodel.predict(X_test)

In [32]:
mae_train = mean_absolute_error(y_train, y_trained)
mae_test = mean_absolute_error(y_test, y_tested)
mse_train = mean_squared_error(y_train, y_trained)
mse_test = mean_squared_error(y_test, y_tested)
rmse_train = np.sqrt(mse_train)
rmse_test = np.sqrt(mse_test)
print("mae_train:", round(mae_train,3)) 
print("mae_test:", round(mae_test,3))
print("mae_train:",round(rmse_train,3))
print("mae_test to 3 DP:",round(rmse_test,3))
print("mae_test not rounded:",rmse_test)

mae_train: 53.742
mae_test: 53.643
mae_train: 95.216
mae_test to 3 DP: 93.64
mae_test not rounded: 93.64046093998027


### Q18 Mean Absolute Error (in three decimal places) for the  training set is 53.742

### Q19 Root Mean Squared Error (in three decimal places) for the training set is 95.216

### Q20 Mean Absolute Error (in three decimal places) for the  testing set is 53.643

### Q21 Root Mean Squared Error (in three decimal places) for the testing set is 93.64

### Q22 No the model did not overfit to training set because RMSE for training is higher than RMSE for testing set

### Ridge Training

In [33]:
from sklearn.linear_model import Ridge 

#ridge
ridge_reg =Ridge(alpha=0.5)
ridge_reg.fit(X_train, y_train)
predicted_ridge_test =ridge_reg.predict(X_test)
predicted_ridge_train =ridge_reg.predict(X_train)

In [34]:
mse_test_ridge = mean_squared_error(y_test, predicted_ridge_test)
rmse_test_ridge = np.sqrt(mse_test_ridge)
print("rmse_test_ridge:", rmse_test_ridge)
print("rmse_test_ridge to 3 DP:", round(rmse_test_ridge,3))

rmse_test_ridge: 93.64051476255226
rmse_test_ridge to 3 DP: 93.641


### Q23 RMSE changed by approximately 0.001 to 93.641 for Ridge regression for testing set. This could be considered insignificant depending on the level of accuracy maintained

### Lasso Training

In [35]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)
predicted_lasso_test = lasso_reg.predict(X_test)
predicted_lasso_train = lasso_reg.predict(X_train)

C:\Users\Anih Nathan\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.136e+06, tolerance: 1.464e+04
  model = cd_fast.enet_coordinate_descent(


In [36]:
def get_weights_df (model, feat, col_name):
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights.reset_index())
    weights_df.columns =['Features',col_name]
    weights_df[col_name].round(3)
    return weights_df

lasso_model_weights = get_weights_df(lasso_reg,X_train, 'Lasso_Model_Weight')

lasso_model_weights
 

Features  Lasso_Model_Weight
0            T9       -2.113836e+01
1            T2       -1.836318e+01
2          RH_2       -1.373723e+01
3         T_out       -1.105957e+01
4          RH_8       -5.778977e+00
5          RH_9       -1.762338e+00
6          RH_7       -1.691936e+00
7            T5       -1.601612e+00
8        RH_out       -1.091186e+00
9            T1       -3.665146e-01
10          rv2        5.951180e-17
11          rv1        3.295809e-02
12  Press_mm_hg        1.702056e-01
13   Visibility        2.025350e-01
14         RH_5        2.575373e-01
15         RH_6        4.114501e-01
16           T7        1.034389e+00
17         RH_4        1.202900e+00
18    Windspeed        2.230566e+00
19           T4        2.788718e+00
20         RH_3        4.802045e+00
21    Tdewpoint        5.688957e+00
22           T6        7.360327e+00
23           T8        9.988261e+00
24         RH_1        1.629967e+01
25           T3        2.583110e+01

### Q24 There are 26 (all) the features with non-zero feature weight for lasso regresion algorithm

In [38]:
mse_test_lasso = mean_squared_error(y_test, predicted_lasso_test)
mse_train_lasso = mean_squared_error(y_train, predicted_lasso_train)
rmse_test_lasso = np.sqrt(mse_test_lasso)
rmse_train_lasso = np.sqrt(mse_train_lasso)
print("rmse_test_lasso:", round(rmse_test_lasso,3))
print("rmse_test_lasso:", rmse_test_lasso)
print("rmse_train_lasso:", round(rmse_train_lasso,3))

rmse_test_lasso: 93.64
rmse_test_lasso: 93.64047772243322
rmse_train_lasso: 95.216


### Q25 RMSE with Lasso Regression on Test set is 93.64